File: core/src/main/java/com/altester/core/service/TestService.java
package com.altester.core.service;

import com.altester.core.dtos.core_service.test.*;
import com.altester.core.exception.*;
import com.altester.core.model.subject.enums.QuestionDifficulty;
import com.altester.core.util.CacheablePage;
import java.security.Principal;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface TestService {

  /**
   * Retrieves a paginated and searchable list of all tests for administrators, with optional
   * filtering by active status.
   *
   * @param pageable Pagination information
   * @param principal The authenticated administrator
   * @param searchQuery Optional search text to filter tests by title
   * @param isActive Optional filter by test activity status (open/closed)
   * @return Paginated list of tests with group information
   * @throws AccessDeniedException if the user is not an administrator
   */
  CacheablePage<TestSummaryDTO> getAllTestsForAdmin(
      Pageable pageable, Principal principal, String searchQuery, Boolean isActive);

  /**
   * Retrieves a paginated and searchable list of tests for a teacher, with optional filtering by
   * active status and edit permission. Only includes tests associated with groups where the teacher
   * is assigned.
   *
   * @param pageable Pagination information
   * @param principal The authenticated teacher
   * @param searchQuery Optional search text to filter tests by title
   * @param isActive Optional filter by test activity status (open/closed)
   * @param allowTeacherEdit Optional filter for tests that allow teacher editing
   * @return Paginated list of tests with group information
   * @throws AccessDeniedException if the user is not a teacher
   */
  CacheablePage<TestSummaryDTO> getTeacherTests(
      Pageable pageable,
      Principal principal,
      String searchQuery,
      Boolean isActive,
      Boolean allowTeacherEdit);

  /**
   * Retrieves summary information for a specific test.
   *
   * @param testId ID of the test to retrieve
   * @param principal The authenticated user
   * @return Test summary with associated groups
   * @throws ResourceNotFoundException if the test doesn't exist
   * @throws AccessDeniedException if the user doesn't have access to the test
   */
  TestSummaryDTO getTestSummary(Long testId, Principal principal);

  /**
   * Retrieves detailed preview information for a specific test, including questions and
   * configuration.
   *
   * @param testId ID of the test to retrieve
   * @param principal The authenticated user
   * @return Detailed test preview
   * @throws ResourceNotFoundException if the test doesn't exist
   * @throws AccessDeniedException if the user doesn't have access to the test
   */
  TestPreviewDTO getTestPreview(Long testId, Principal principal);

  /**
   * Retrieves a paginated and searchable list of tests associated with a specific subject. For
   * teachers, only includes tests in groups they teach.
   *
   * @param subjectId ID of the subject
   * @param principal The authenticated user
   * @param searchQuery Optional search text to filter tests by title
   * @param isActive Optional filter by test activity status
   * @param pageable Pagination information
   * @return Paginated list of tests with group information
   * @throws ResourceNotFoundException if the subject doesn't exist
   */
  CacheablePage<TestSummaryDTO> getTestsBySubject(
      Long subjectId, Principal principal, String searchQuery, Boolean isActive, Pageable pageable);

  /**
   * Retrieves a paginated and searchable list of tests associated with a specific group.
   *
   * @param groupId ID of the group
   * @param principal The authenticated user
   * @param searchQuery Optional search text to filter tests by title
   * @param isActive Optional filter by test activity status
   * @param pageable Pagination information
   * @return Paginated list of tests with group information
   * @throws ResourceNotFoundException if the group doesn't exist
   * @throws AccessDeniedException if the user doesn't have access to the group
   */
  CacheablePage<TestSummaryDTO> getTestsByGroup(
      Long groupId, Principal principal, String searchQuery, Boolean isActive, Pageable pageable);

  /**
   * Retrieves a paginated list of questions for a specific test, with optional filtering by
   * difficulty level. This method allows administrators and teachers to view all questions
   * associated with a test.
   *
   * @param testId ID of the test to retrieve questions for
   * @param principal The authenticated user
   * @param pageable Pagination information
   * @param difficulty Optional filter to show only questions of a specific difficulty level
   * @return Paginated list of questions for the test
   * @throws ResourceNotFoundException if the test doesn't exist
   * @throws AccessDeniedException if the user doesn't have access to the test
   */
  Page<QuestionDTO> getTestQuestions(
      Long testId, Principal principal, Pageable pageable, QuestionDifficulty difficulty);

  /**
   * Generates a randomized test preview for students with the configured number of questions at
   * each difficulty level. The questions are randomly selected based on the test configuration and
   * shuffled.
   *
   * @param testId ID of the test to generate a preview for
   * @param principal The authenticated student
   * @param pageable Pagination information
   * @return Paginated list of randomly selected questions for the student's test attempt
   * @throws ResourceNotFoundException if the test doesn't exist
   * @throws AccessDeniedException if the user doesn't have access to the test
   */
  Page<QuestionDTO> getStudentTestPreview(Long testId, Principal principal, Pageable pageable);
}

File: frontend/src/services/TestService.js
import {
  BASE_API_URL,
  createAuthAxios,
  handleApiError,
  buildUrlWithParams,
} from './apiUtils';

const API_URL = BASE_API_URL;

export const TestService = {
  getAllTests: async (
    page = 0,
    size = 10,
    sort = 'id,desc',
    searchQuery = '',
    isActive = null
  ) => {
    try {
      const axiosInstance = createAuthAxios();

      const params = {
        page,
        size,
        sort,
        searchQuery,
        isActive,
      };

      const url = buildUrlWithParams(`${API_URL}/admin/tests`, params);
      const response = await axiosInstance.get(url);
      return response.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  getTeacherTests: async (
    page = 0,
    size = 10,
    sort = 'id,desc',
    searchQuery = '',
    isActive = null,
    allowTeacherEdit = null
  ) => {
    try {
      const axiosInstance = createAuthAxios();

      const params = {
        page,
        size,
        sort,
        searchQuery,
        isActive,
        allowTeacherEdit,
      };

      const url = buildUrlWithParams(`${API_URL}/teacher/tests/my`, params);
      const response = await axiosInstance.get(url);
      return response.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  getTestPreview: async testId => {
    try {
      const axiosInstance = createAuthAxios();
      const response = await axiosInstance.get(
        `${API_URL}/teacher/tests/${testId}/preview`
      );
      return response.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  getTestSummary: async testId => {
    try {
      const axiosInstance = createAuthAxios();
      const response = await axiosInstance.get(
        `${API_URL}/teacher/tests/${testId}/summary`
      );
      return response.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  createTest: async testData => {
    try {
      if (testData.groupIds) {
        testData.groupIds = [...new Set(testData.groupIds)];
        testData.assignmentType = 'groups';
      }

      const axiosInstance = createAuthAxios();
      const response = await axiosInstance.post(
        `${API_URL}/teacher/tests`,
        testData
      );
      return response.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  updateTest: async (testId, testData) => {
    try {
      if (testData.groupIds) {
        testData.groupIds = [...new Set(testData.groupIds)];
        testData.assignmentType = 'groups';
      }

      const axiosInstance = createAuthAxios();
      const response = await axiosInstance.put(
        `${API_URL}/teacher/tests/${testId}`,
        testData
      );
      return response.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  deleteTest: async testId => {
    try {
      const axiosInstance = createAuthAxios();
      const response = await axiosInstance.delete(
        `${API_URL}/teacher/tests/${testId}`
      );
      return response.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  toggleTestActivity: async testId => {
    try {
      const axiosInstance = createAuthAxios();
      const response = await axiosInstance.put(
        `${API_URL}/teacher/tests/${testId}/activity`
      );
      return response.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  toggleTeacherEditPermission: async testId => {
    try {
      const axiosInstance = createAuthAxios();
      const response = await axiosInstance.put(
        `${API_URL}/admin/tests/${testId}/teacher-edit`
      );
      return response.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  toggleTestAiEvaluation: async testId => {
    try {
      const axiosInstance = createAuthAxios();
      const response = await axiosInstance.put(
        `${API_URL}/tests/${testId}/evaluation`
      );
      return response.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  getTestsBySubject: async (
    subjectId,
    searchQuery = '',
    isActive = null,
    page = 0,
    size = 10
  ) => {
    try {
      const axiosInstance = createAuthAxios();

      const params = {
        page,
        size,
        searchQuery,
        isActive,
      };

      const url = buildUrlWithParams(
        `${API_URL}/teacher/tests/subject/${subjectId}`,
        params
      );
      const response = await axiosInstance.get(url);
      return response.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  getTestsByGroup: async (
    groupId,
    searchQuery = '',
    isActive = null,
    page = 0,
    size = 10
  ) => {
    try {
      const axiosInstance = createAuthAxios();

      const params = {
        page,
        size,
        searchQuery,
        isActive,
      };

      const url = buildUrlWithParams(
        `${API_URL}/teacher/tests/group/${groupId}`,
        params
      );
      const response = await axiosInstance.get(url);
      return response.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  getSubjectsForTest: async (page = 0, searchQuery = '') => {
    try {
      const axiosInstance = createAuthAxios();

      const params = {
        page,
        size: 12,
        searchQuery,
      };

      const url = buildUrlWithParams(`${API_URL}/admin/subject/all`, params);
      const response = await axiosInstance.get(url);
      return response.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  getGroupsForTest: async (
    page = 0,
    searchQuery = '',
    activityFilter = 'active'
  ) => {
    try {
      const axiosInstance = createAuthAxios();

      const params = {
        page,
        size: 12,
        searchQuery,
        activityFilter,
      };

      const url = buildUrlWithParams(`${API_URL}/admin/group/all`, params);
      const response = await axiosInstance.get(url);
      return response.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  getSubjectDetails: async subjectId => {
    try {
      const axiosInstance = createAuthAxios();
      const response = await axiosInstance.get(
        `${API_URL}/admin/subject/${subjectId}`
      );
      return response.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  getTestQuestions: async (testId, page = 0, size = 20, difficulty = null) => {
    try {
      const axiosInstance = createAuthAxios();

      const params = {
        page,
        size,
        difficulty,
      };
      const url = buildUrlWithParams(
        `${API_URL}/teacher/tests/${testId}/questions`,
        params
      );
      const response = await axiosInstance.get(url);
      return response.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  getStudentTestPreview: async (testId, page = 0, size = 20) => {
    try {
      const axiosInstance = createAuthAxios();

      const params = {
        page,
        size,
      };

      const url = buildUrlWithParams(
        `${API_URL}/teacher/tests/${testId}/student-preview`,
        params
      );
      const response = await axiosInstance.get(url);
      return response.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },
};
